karapace/schema_registry.py

```diff
@@ -393,17 +392,15 @@ async def write_new_schema_local(
             )
             return schema_id
 
-    def get_versions(self, schema_id: int, *, include_deleted: bool = False) -> List[SubjectData]:
+    def get_versions(self, schema_id: int, *, include_deleted: bool = False) -> List[Dict[Subject, ResolvedVersion]]:
         subject_versions = []
-        with self.schema_reader.id_lock:
-            for subject, val in self.schema_reader.subjects.items():
-                if self.schema_reader.get_schemas(subject, include_deleted=include_deleted) and "schemas" in val:
-                    schemas = val["schemas"]
-                    for version, schema in schemas.items():
-                        if int(schema["id"]) == schema_id and (include_deleted or not schema.get("deleted", False)):
-                            subject_versions.append({"subject": subject, "version": int(version)})
+        schema_versions = self.database.find_schema_versions_by_schema_id(
+            schema_id=schema_id, include_deleted=include_deleted
+        )
+        for schema_version in schema_versions:
+            subject_versions.append({"subject": schema_version.subject, "version": schema_version.version})
         subject_versions = sorted(subject_versions, key=lambda s: (s["subject"], s["version"]))
-        return subject_versions
+        return cast(List[Dict[Subject, ResolvedVersion]], subject_versions)  # Keep mypy happy
```

Comment:
question: What is mypy failing to infer here, could it be fixed with explicit hint on `subject_versions`?

Reply:
The types were off, corrected.